Src/Support/GoogleApis.Tests/Apis/Upload/ImprovedResumableUploadTest.cs

```diff
@@ -0,0 +1,966 @@
+                    }
+                } while (_httpListener == null);
+                _httpTask = RunServer();
+            }
+
+            private readonly HttpListener _httpListener;
+            private readonly Task _httpTask;
+
+            public string HttpPrefix { get; }
+
+            private async Task RunServer()
+            {
+                while (_httpListener.IsListening)
+                {
+                    var context = await _httpListener.GetContextAsync();
+                    var response = context.Response;
+                    if (context.Request.Url.AbsolutePath.EndsWith("/[Quit]"))
+                    {
+                        response.Close();
+                        _httpListener.Stop();
+                    }
+                    else
+                    {
+                        response.ContentType = "text/plain";
+                        IEnumerable<byte> body;
+                        try
+                        {
+                            body = await HandleCall(context.Request, response);
+                            var bodyBytes = body?.ToArray() ?? new byte[0];
+                            await response.OutputStream.WriteAsync(bodyBytes, 0, bodyBytes.Length);
+                        }
+                        catch (HttpListenerException) { }
+                        finally
+                        {
+                            try
+                            {
+                                response.Close();
+                            }
+                            catch (Exception) { }
+                        }
+                    }
+                }
+            }
+
+            public abstract class Handler : IDisposable
+            {
+                private static int handlerId = 0;
+
+                private TestServer _server;
+
+                public Handler(TestServer server)
+                {
+                    _server = server;
+                    Id = Interlocked.Increment(ref handlerId).ToString();
+                    _server.RegisterHandler(this);
+                }
+
+                public string Id { get; }
+                public string HttpPrefix => $"{_server.HttpPrefix}{Id}/";
+
+                public string RemovePrefix(string s)
+                {
+                    var prefix = $"/{Id}/";
+                    if (s.StartsWith(prefix))
+                    {
+                        return s.Substring(prefix.Length);
+                    }
+                    throw new InvalidOperationException("Doesn't start with prefix");
+                }
+
+                public List<RequestInfo> Requests { get; } = new List<RequestInfo>();
+
+                public Task<IEnumerable<byte>> HandleCall0(
+                    HttpListenerRequest request, HttpListenerResponse response)
+                {
+                    Requests.Add(new RequestInfo(request));
+                    return HandleCall(request, response);
```

Comment:
For future diagnostics, I wonder whether it would be worth changing RequestInfo into RequestResponseInfo, so we can see requests and responses at the same time.